Offline mode: Show conflict resolution dialog from post list

momo-ozawa opened the pull request:
Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/22583

## Description
* Adds an entry point to the conflict resolution dialog from the post list
  * If a post has failed with a conflict error in the post list, tapping the cell will open the conflict resolution dialog
  * After resolving a post conflict, tapping the cell will open the editor

## How to test

### Test 1.1 - Update published post 

_Verify picking both "Current version" and "Another version" in the conflict resolution dialog works as expected._

- Open a published post in the **web**
- Open the same published post in the **app**
- Makes changes to the post in the **web**, and tap "Update"
- Make different changes to the post in the **app**, and tap "Update"
- Go through the conflict resolution flow and pick a version
- ✅ Verify the expected version is displayed in the **app**
- Refresh the post in the **web**
- ✅ Verify the expected version is displayed in the **web**

### Test 2.1 - Save draft post

_Verify picking both "Current version" and "Another version" in the conflict resolution dialog works as expected._

- Open a draft post in the **web**
- Open the same draft post in the **app**
- Makes changes to the post in the **web**, and tap "Save draft"
- Make different changes to the post in the **app**, and tap back then "Save draft"
- ✅ Verify the post cell displays a warning icon
- Tap on the cell
- Go through the conflict resolution flow and pick a version
- ✅ Verify the expected version is displayed in the **app**
- ✅ Verify the post cell no longer displays a warning icon
- ✅ Verify tapping on the post cell opens up the editor
- Refresh the post in the **web**
- ✅ Verify the expected version is displayed in the **web**

### Test 2.2.1 - Publish draft post from context menu

_Verify picking both "Current version" and "Another version" in the conflict resolution dialog works as expected._

- Open a draft post in the **web**
- Open the same draft post in the **app**
- Makes changes to the post in the **web**, and tap "Save draft"
- Make different changes to the post in the **app**, and tap back then "Save draft"
- ✅ Verify the post cell displays a warning icon
- Tap on the context menu, and tap "Publish"
- Tap "Publish" in the pre-publishing sheet
- Go through the conflict resolution flow and pick a version
- Tap "Publish" in the pre-publishing sheet again
-  ✅ Verify the expected version is in the published tab in the **app**
-  ✅ Verify the post cell no longer displays a warning icon
-  ✅ Verify tapping on the post cell opens up the editor
-  Refresh the post in the **web**
-  ✅ Verify the expected version is displayed in the **web**


### Test 2.2.2 - Publish draft post from editor

_Verify picking both "Current version" and "Another version" in the conflict resolution dialog works as expected._

- Open a draft post in the **web**
- Open the same draft post in the **app**
- Makes changes to the post in the **web**, and tap "Save draft"
- Make different changes to the post in the **app**, and tap "Publish"
- Tap "Publish" in the pre-publishing sheet
- Go through the conflict resolution flow and pick a version
- Tap "Publish" in the pre-publishing sheet again
- ✅ Verify the expected version is in the published tab in the **app**
- ✅ Verify the post cell no longer displays a warning icon
- ✅ Verify tapping on the post cell opens up the editor
- Refresh the post in the **web**
- ✅ Verify the expected version is displayed in the **web**


## Regression Notes
1. Potential unintended areas of impact
post conflict

15. What I did to test those areas of impact (or what existing automated tests I relied on)
manual

16. What automated tests I added (or what prevented me from doing so)
n/a

PR submission checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

Commit:
show conflict resolution dialog from post list

Commit:
subscribe to resolve conflict notif in PostEditor

dangermattic commented:
<!--
  0 Errors
  1 Warning: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

Commit:
resolve conflict on post.original()

Commit:
use postCoordinatorDidUpdate notif to update editor

Commit:
fire notifications using pre-existing method

Commit:
subscribe to postCoordinatorDidUpdate notif in prepublishing vc

Commit:
remove the postConflictResolved notification

conflict resolution now relies on the postCoordinatorDidUpdate notification

Commit:
rely on postID to identify updated post

Commit:
fix swiftlint error

Commit:
use wpAssert

kean commented on WordPress/Classes/Services/PostCoordinator.swift:
```diff
@@ -591,6 +601,13 @@ class PostCoordinator: NSObject {
     }
 
+    private func shouldScheduleRetry(for error: Error) -> Bool {
+        if error is PostRepository.PostSaveError {
```
I would suggest checking for a specific case and making it so that you need to opt-out of retries if you add new error cases.

For `.deleted` (aka not found) this branch is unreachable.

momo-ozawa replied:
Good point - fixed! https://github.com/wordpress-mobile/WordPress-iOS/pull/22992/commits/02cea1b147580af4e1fc3d30a432703a2a976068

kean commented on WordPress/Classes/Services/PostCoordinator.swift:
```diff
@@ -479,6 +485,8 @@ class PostCoordinator: NSObject {
 
     private func startSync(for post: AbstractPost) {
         guard let revision = post.getLatestRevisionNeedingSync() else {
+            let worker = getWorker(for: post)
```
If you change the publicly visible upload state, it should also send `.postDidUpdateNotification`.

momo-ozawa replied:
Done https://github.com/wordpress-mobile/WordPress-iOS/pull/22992/commits/522b57160de80a3823edbc177ed1b6871e1695d4

kean commented on WordPress/Classes/Services/PostCoordinator.swift:
```diff
@@ -284,6 +285,11 @@ class PostCoordinator: NSObject {
     }
 
+    func didResolveConflict(for post: AbstractPost) {
+        workers[post.objectID]?.error = nil
```
Is this code still needed?

momo-ozawa replied:
I revised this to call `startSync(for:)` because we still need to clear the error - is that what you mean? https://github.com/wordpress-mobile/WordPress-iOS/pull/22992/commits/30bbdd72a9faa1c8910e44a40a8e1f4d490b3b12

kean commented on WordPress/Classes/ViewRelated/Post/PostEditor.swift:
```diff
@@ -181,9 +198,23 @@ extension PostEditor where Self: UIViewController {
     }
+
+    private func postCoordinatorDidUpdate(_ notification: Foundation.Notification) {
+        guard let updatedObjects = (notification.userInfo?[NSUpdatedObjectsKey] as? Set<NSManagedObject>) else {
```
The `postCoordinatorDidUpdate` is called frequently during the upload. For example, you can open an unsynced draft post and continue making changes while the previous revision are being uploaded in the background. 

I would suggest adding the previous notification or introducing a new more generic notification for when the latest revision gets destroyed by the coordinator, otherwise the editor will keep creating revisions when it doesn't need to, and it will break the `hasChanges` mechanism.

momo-ozawa replied:
Done! https://github.com/wordpress-mobile/WordPress-iOS/pull/22992/commits/d5dfff314e0c4fe437184121a3292b8836cccc91

kean commented on WordPress/Classes/ViewRelated/Post/PostListViewController.swift:
```diff
@@ -196,6 +196,15 @@ final class PostListViewController: AbstractPostListViewController, UIViewContro
             return
         }
 
+        if let error = PostCoordinator.shared.syncError(for: post.original()),
```
Please move it to `PostListEditorPresenter` and also add it to the `PagetListEditorPresenter` and test that it works in search.

momo-ozawa replied:
Fixed! https://github.com/wordpress-mobile/WordPress-iOS/pull/22992/commits/e3a7a868f66185658fb4e5bd14ea74dfc949bc6f

This is now working as expected for post list, page list, and search

kean commented on WordPress/Classes/ViewRelated/Post/Prepublishing/PrepublishingViewController.swift:
```diff
@@ -680,6 +691,22 @@ private extension PrepublishingOption {
+// MARK: - Notifications
+extension PrepublishingViewController {
+    private func postCoordinatorDidUpdate(_ notification: Foundation.Notification) {
+        guard let updatedObjects = (notification.userInfo?[NSUpdatedObjectsKey] as? Set<NSManagedObject>) else {
```
Could you please clarify the intention behind adding this observer?

momo-ozawa replied:
Added a clarifying comment b75e13d

This handles the case when you try to publish a draft post with a conflict.

kean commented:
Failing test scenarios:

**Test 3.1**

- Set a breakpoint on `/rest/v1.2/sites/*/posts/*
- Save a change to a draft
- **Verify** that it hits the breakpoint
- Reopen the draft and make more edits
- Release the breakpoint
- Tap "back" and **verify** that the editor suggest to save edits and can save them

Observed result:

The app hits the breakpoint (it should not try to create a new local revision in this scenario):

```
assert(post.latest() == post, "Must be opened with the latest verison of the post")
```

**Test 3.2**

Precondition: a draft post has a conflict.

- Search for a post with a conflict
- Tap the cell with a post
- **Verify** that the conflict resolution dialog is shown

**Test 3.3**

Precondition: a draft **page** has a conflict.

- Tap the cell with a page
- **Verify** that the conflict resolution dialog is shown

Commit:
check for specific error cases when opting out of retry

Commit:
send notification after resetting error state

Commit:
Revert "remove the postConflictResolved notification"

This reverts commit b5a97ce988b5b2981f446c9a949ae82199983600.

Commit:
use postConflictResolved notif to update editor

Intentionally not using postCoordinatorDidUpdate to avoid breaking the sync mechanism.

The postCoordinatorDidUpdate is called frequently during the upload. If we use the postCoordinatorDidUpdate here, the editor will keep creating revisions when it doesn't need to, and it will break the hasChanges mechanism.

Commit:
show resolve conflict dialog for post list, page list, and search

Commit:
clear conflict error by calling startSync

Commit:
add clarifying comment for resolve conflict notification
